File: audio.pl
use POSIX;
$tau = 2 * 3.1415;
sub fmmod {
  my ($t, $fm, $fc, $fd) = @_;
  return sin($tau * $fc * $t + $fd * cos($t * $tau * $fm)/$fm);
}
sub p1 {
  my ($t) = @_;
  return (1 - $t)**2;
}
sub p2 {
  my ($t) = @_;
  return (2 - 2**$t);
}
my @c = ();
srand(666);
for(my $i = 0; $i < 16; $i++){
  push @c, (261, 293, 329, 349, 392, 440, 493)[int(rand(7))];
}
my $r = 0;
sub dsp {
  my $t = 2*(pop)/22050.0;
    my $blen = 3;
  my $prog = fmod($t * $blen, 1);
  my $bar = int($t * $blen);
  my $decay = (1-(1-$prog)**16) * p1($prog);
  my $decay2 = (1-(1-$prog)**16) * p2($prog);
    my $codes = $c[$bar % 16]*2;
    my $kick = 0;
  my $melody = 0;
    if($bar < 80) {
    $melody += fmmod($t, 5, $codes, $bar*4);
  }
    if($bar == 81) {
    return chr 4;
  }
  if($bar < 64) {
    if($bar % 4 == 0 && $bar >= 16){
      $kick += fmmod($t, 50, 60, 10);
      if(rand() < 0.05 ) {
        $r = rand()-0.5;
      }
      $kick += $r/2;
    }
    if($bar % 4 == 2 && $bar >= 32){
      if(rand() < (0.2 + ($bar % 8 == 0)/3 ) ) {
        $r = rand()-0.5;
      }
      $kick += $r/2;
    }
    if($bar % 2 == 1 && $bar >= 48){
      if(rand() < (0.8 + ($bar % 8 == 0) ) ) {
        $r = rand()-0.5;
      }
      $kick += $r/3;
    }
  }
  return $melody/4*$decay + $kick/2*$decay2;
}

print "Now Playing: Crashing Ducks\n";

my ($read, $write) = POSIX::pipe();
my $pid=fork();
if($pid != 0){
  POSIX::close( $write );
  dup2($read, STDIN);
  exec("/usr/bin/aplay -q -c 1 -r 22050 -f S16_LE");
} else {
  POSIX::close($read);

  my @buf = map{0}(1..1024);
  for(my$t=0;;$t++){
    $g = $t % 1024;
    if($t % 1024 == 0){
      my $r = pack 's<[1024]', @buf;
      POSIX::write($write, $r, 1024*2);
    }
    $dsp_g = dsp($t);
    if($dsp_g eq chr 4){
      exit 0;
    }
    $buf[$g] = $dsp_g * 32768;
  }
}

File: audio.py
#!/usr/bin/env python2.7
import posix as px
import sys
import array
from math import *
import random

rate = 22050
time = 0

def p2m(pitch):
	global rate
	return (pi*2*pitch)/rate

def sec2samp(sec):
	return (1.0*sec)*rate

def samp2sec(time):
	return (1.0*time)/rate

samplelength = int(sec2samp(0.005))+1
samples = [0]*samplelength

class Rosenberg:
	def __init__(self, ts, cps, p1, p2, p3):
		self.time = int(0)
		self.sum = float(0)
		self.ts = float(ts)
		self.cps = float(cps)
		self.p1 = float(p1)
		self.p2 = float(p2)
		self.p3 = float(p3)

	def step(self):
		self.time += 1
		t = self.time
		t %= int(sec2samp(1.0/self.cps))
		s = samp2sec(t)/self.ts
		diff = 0
		if (s < 1):
			diff += self.p1*pow(s, self.p1-1)
			diff -= self.p2*pow(s, self.p2-1)
			diff += pow(s, self.p3)
			diff -= 1.0/(self.p3 + 1.0)
		self.sum += diff/sec2samp(self.ts)
		self.sum *= 0.995 #correct any bias toward 0
		return self.sum

rosen = Rosenberg(0.007, 100, 2.0, 3.0, 60.0)

def convolve(samples, samplelength):
	# return samples[0]
	mysum = 0.0
	mymax = 0.0
	for i in xrange(0, samplelength, 4):
		sec = samp2sec(i)*1+0.001
		mult = sin(sec*1)/sec
		# print max(0.0,1.0-sec)
		mysum += samples[i] * mult
		mymax += mult

	return mysum/mymax

def gensample():
	global samples
	global samplelength
	global rosen
	newsample = rosen.step()
	samples.pop()
	samples.insert(0,newsample)

	# outsample = 0
	# for x in xrange(0, 100):
	# 	outsample += samples[x]
	# outsample /= 10.0
	return convolve(samples, samplelength)*10

def genbuf(buf, bufsize):
	for i in xrange(0, bufsize):
		buf[i] = 0.0
	lamb = 0.1
	tl_len = 300
	tbuff = []
	std = sqrt(1.0/(tl_len*lamb))
	
	for i in range(0, tl_len):
		scale = (1.0*i)/tl_len
		weight = (1.0-cos(scale*2*pi))
		result = weight*sin(i*p2m(400))*0.2
		tbuff.append(result)

	if True:
		pulselen = 2500
		echolen = 5
		for off in xrange(0, echolen):
			scale2 = pow(1-(off*1.0)/echolen,3)
			for i in xrange(0,pulselen):
				scale = pow((1.0*i)/pulselen,2)
				weight = (1.0-cos(scale*2*pi))
				buf[off*pulselen*8+i] += scale2*weight*sin(i*p2m(1058))/4.0

	try:
		tk = 0
		while tk+tl_len < bufsize:
			rando = random.choice([-std,std])
			for i in range(0, tl_len):
				buf[tk+i] += tbuff[i]*rando

			rand = log(1-random.uniform(0, 1))/lamb
			tk -= rand
			tk = int(tk)
			# print tk
	except:
		return

def main():
	read, write = px.pipe()
	pid = px.fork()
	if (pid != 0):
		px.close(read)

		bufsize = 1024*64*2
		buf = array.array('f', [0]*bufsize)

		while 1:
			try:
				genbuf(buf, bufsize)
				px.write(write, buf.tostring())
			except KeyboardInterrupt:
				exit(0)
	else:
		px.close(2)
		px.close(write)
		px.dup2(read, 0)
		aplay_args = ("/usr/bin/aplay -q -c 1 -r %d -f FLOAT_LE" % rate).split(" ")
		px.execv(aplay_args[0], aplay_args)

main()